core/embed/trezorhal/rdi.c

```diff
@@ -70,46 +91,54 @@ __attribute__((target("thumb"))) void rdi_handler(uint32_t uw_tick) {
 
     asm volatile(
         "ldr r0, %0;"  // r0 = delay
-        "add r0, $3;"  // r0 += 3
+        "add r0, $6;"  // r0 += 6
         "loop:"
         "subs r0, $3;"  // r0 -= 3
         "bhs loop;"     // if (r0 >= 3): goto loop
-        // loop ((delay // 3) + 1) times
-        // one extra loop ensures that branch predictor learns the loop
+        // loop ((delay // 3) + 2) times
+        // two extra loops ensures that branch predictor learns the loop
         // every loop takes 3 ticks
         // r0 == (delay % 3) - 3
-        "lsl r0, $1;"      // r0 *= 2
-        "add r0, $4;"      // r0 += 4
-        "rsb r0, r0, $0;"  // r0 = -r0
-        // r0 = 2 if (delay % 3 == 0) else 0 if (delay % 3 == 1) else -2 if
-        // (delay % 3 == 2)
-        "add pc, r0;"  // jump (r0 + 2)/2 instructions ahead
-        // jump here if (delay % 3 == 2)
-        "nop;"  // wait one tick
-        // jump here if (delay % 3 == 1)
-        "nop;"  // wait one tick
-        // jump here if (delay % 3 == 0)
+        "add r0, $3;"  // r0 += 3
+        // r0 == delay % 3
+        "and r0, r0, $3;"  // r0 %= 3, make sure that 0 <= r0 < 4
+        "ldr r1, =table;"  // r1 = &table
+        "tbb [r1, r0];"    // jump 2*r1[r0] bytes forward, that is goto wait_r0
+        "base:"
+        "b wait_2;"  // goto wait_2, make sure that we won't execute the table
+        "table:"     // table of branch lengths
+        ".byte (wait_0 - base)/2;"
+        ".byte (wait_1 - base)/2;"
+        ".byte (wait_2 - base)/2;"
+        ".byte (wait_2 - base)/2;"  // next instruction must be 2-byte aligned
+        "wait_2:"
+        "mov r0, r0;"  // wait one tick
+        "wait_1:"
+        "mov r0, r0;"  // wait one tick
+        "wait_0:"
         :
         : "m"(delay)
-        : "r0");  // wait (24 + delay) ticks
+        : "r0", "r1");  // wait (44 + delay) ticks
```

Comment:
BTW, I would put this comment at the top rather than the bottom, because by the time the reader gets here, they will have already spent time figuring out what the code does.

Reply:
Done.